### PR TITLE
Change rule for submodules update (fix issue #123)

### PR DESCRIPTION
**Pull request by stweil**
Signed-off-by: Stefan Weil <sw@weilnetz.de>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -99,6 +99,7 @@ help: ;	@eval "$$HELP"
 #   unless they are already up-to-date
 # - then updates the time stamps of the module directories
 #   so each directory can be used as a dependency
+$(OCRD_MODULES): modules
```
Comment (bertsky) [inline Makefile:102]: This introduces a circular dependency, because we already have `modules: $(OCRD_MODULES)`

**Comment by stweil** [comment]
@bertsky, @kba, the PR now passes CI and should be ready for merging.

**Comment by stweil** [comment]
> it sacrifices the ability to build modules independent of each other

Building modules was not changed. The pull request only changes the way how the submodules are synchronized and updated. This results in a changed behaviour for some less common scenarios: running `make ocrd` not only synchronizes and updates the `core` submodule, but also all other submodules according to `OCRD_MODULES`. By setting that macro it is possible to enforce the old behaviour.

**Comment by bertsky** [comment]
> > it sacrifices the ability to build modules independent of each other
> 
> Building modules was not changed. The pull request only changes the way how the submodules are synchronized and updated.

It throws them all together into a single rule (named `modules`). This can slow down building severely, especially for modules with large tails. Parallel build should start using the CPU as soon as the network transfer is ready for the first modules. (Not all submodule actions need a lock file and thus have to be synchronized.)

> This results in a changed behaviour for some less common scenarios: running `make ocrd` for example not only synchronizes and updates the `core` submodule, but also all other submodules according to `OCRD_MODULES`.

True! That's another minus for this approach IMO.

File: Makefile [hunk 2]
```diff
@@ -93,20 +93,30 @@ export HELP
...
-#   unless it is already up-to-date
-# - then updates the time stamp of the module directory
-#   so the directory can be used as a dependency
-modules: $(OCRD_MODULES)
+# - updates the submodules to the revision registered here,
+#   unless they are already up-to-date
+# - then updates the time stamps of the module directories
+#   so each directory can be used as a dependency
+$(OCRD_MODULES): modules
+RECURSIVE_MODULES = ocrd_fileformat ocrd_olena opencv-python
```
Comment (bertsky) [inline Makefile:103]: Defining this in a central place undermines our principle of encapsulating all modules' knowledge together. You could easily use `RECURSIVE_MODULES += ocrd_fileformat` (asf.) style.

File: Makefile [hunk 3]
```diff
@@ -93,20 +93,30 @@ export HELP
...
+	    if git submodule status $(GIT_RECURSIVE) $$module | grep -qv '^ '; then \
+		git submodule update --init $(GIT_RECURSIVE) $$module && \
+		touch $$module; \
+	    fi; \
+	done
+	for module in $(filter $(RECURSIVE_MODULES),$(OCRD_MODULES)); do \
+	    if git submodule status --recursive $$module | grep -qv '^ '; then \
+		git submodule update --init --recursive $$module && \
+		touch $$module; \
+	    fi; \
```
Comment (bertsky) [inline Makefile:118]: Why split the recipe into two parts, copying all commands? Copying always introduces the danger of someone forgetting to update the other accordingly.

You can easily use some `$(if ...)` within the command to discern your `RECURSIVE_MODULES` from the others.

File: Makefile (hunk 1 above)
Comment (kba) [reply Makefile:102]: Eek, that cannot be good. @stweil?

Comment (bertsky) [reply Makefile:102]: @stweil already avoided that by also removing the `modules: $(OCRD_MODULES)` dependency and force-pushing. But I stand by my other criticism (both on how this is done and on the general approach) and still recommend for #125

**Comment by stweil** [comment]
This pull request was superseded by the merged pull request #125, so it can be closed.

**Comment by kba** [comment]
thanks @stweil, I appreciate your continued contributions!

**Pull request closed without merging**